### PR TITLE
V22F-274-connect-to-database

### PR DESCRIPTION
**Pull request by sjoeman1**
Made a connection to the external database.
To test there are files needed with confidential information, ask @sjoeman1 for them.
They need to be added or updated in the data folder, and the certificates need to be placed in a "certificates" folder in data.

### CODE REVIEW

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp [hunk 1]
```diff
@@ -73,23 +73,64 @@ void SQLDatabaseStorage::StoreData(const std::experimental::filesystem::path& p_
...
-/// @param p_port Port the database is located on on the host.
-/// @param p_username Username to connect with to the database.
-/// @param p_password Password to connect with to the database.
-/// @param p_schemaName Name of the database schema to use.
-/// @return returns true if connection to database has been made, false otherwise
+/// @brief                          gets the keys for secure database connection in the data/certificates folder
+///                                 and adds them to the connection properties
+///                                 name of the keys are set in the database_encryption_settings.txt file
+/// @param p_connectionProperties   SQL connection properties to which the keys are added.
+void SQLDatabaseStorage::GetKeys(sql::ConnectOptionsMap p_connectionProperties)
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp:80]: I would rename this function to PutKeys to make it more intuitive what it does.

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp [hunk 2]
```diff
@@ -73,23 +73,64 @@ void SQLDatabaseStorage::StoreData(const std::experimental::filesystem::path& p_
...
+}
+
+/// @brief                  Connect to the specified database.
+///                         Initialise the database with the proper tables if they don't exist yet.
+/// @param p_hostName       Hostname of the database to connect to. Should be a TCP-IP address.
+/// @param p_port           Port the database is located on on the host.
+/// @param p_username       Username to connect with to the database.
+/// @param p_password       Password to connect with to the database.
+/// @param p_schemaName     Name of the database schema to use.
+/// @param p_useEncryption  if encryption is used, "true" if used, otherwise its not used.
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp:119]: *it's/it is

File: speed-dreams/source-2.2.3/data/database_connection_settings.txt [hunk 3]
```diff
@@ -2,4 +2,5 @@
 3306
 root
 PASSWORD
-sda_schema
\ No newline at end of file
+sda_schema
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/data/database_connection_settings.txt:5]: is this supposed to be in here, if it shouldn't we need to change the password of the server

**Review by KungFuDonkey — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp [hunk 4]
```diff
@@ -100,13 +141,19 @@ bool SQLDatabaseStorage::OpenDatabase(
     connection_properties["CLIENT_MULTI_STATEMENTS"] = false;
     connection_properties["sslEnforce"] = true;
 
+    if (p_useEncryption == "true")
```
Comment (Sindarin27) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp:144]: Might want to add to_lower to ensure putting "True" or "TRUE" or similar spellings work as well

File: speed-dreams/source-2.2.3/data/database_connection_settings.txt (hunk 3 above)
Comment (sjoeman1) [reply speed-dreams/source-2.2.3/data/database_connection_settings.txt:5]: that is the database name, should be in there

**Review by Sindarin27 — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/data/database_connection_settings.txt (hunk 3 above)
Comment (sjoeman1) [reply speed-dreams/source-2.2.3/data/database_connection_settings.txt:5]: This is also for the local database
the settingsfile for the remote is different
but not on github

**Review by Sindarin27 — APPROVED** [review]
Approved on code standards. If someone else could confirm this PR works when connecting to the actual database, that would be great.

**Review by KungFuDonkey — CHANGES_REQUESTED** [review]
clang format needs to be done again

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp [hunk 5]
```diff
@@ -73,19 +73,57 @@ void SQLDatabaseStorage::StoreData(const std::experimental::filesystem::path& p_
...
+    std::string pubName;
+    std::string privName;
+    READ_INPUT(ifstream, caName)
+    READ_INPUT(ifstream, pubName)
+    READ_INPUT(ifstream, privName)
+
+    std::string certificatesPath(ROOT_FOLDER "\\data\\certificates");
+    std::string caFile(caName);
+    std::string pubFile(pubName);
+    std::string privFile(privName);
```
Comment (noutheesink) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp:101]: You don't have to copy these string

**Review by noutheesink — CHANGES_REQUESTED** [review]

**Review by noutheesink — APPROVED** [review]

**Pull request merged**